File: backend/app/services/order_service.py
from datetime import datetime
from random import randint
from fastapi import HTTPException, status
from sqlalchemy.orm import Session
from app.models.order import Order, OrderItem
from app.repositories import cart_repository, menu_repository, order_repository
from app.schemas.order import OrderCreate

DELIVERY_FEE = 3.0
VALID_PAYMENT_METHODS = {"CashOnDelivery", "Card", "Wallet"}
CANCELABLE_STATUSES = {"Pending", "Confirmed"}


def _order_response(order: Order) -> Order:
    if getattr(order, "user", None):
        order.customer_name = order.user.full_name
    return order


def _generate_order_number() -> str:
    return f"ORD-{datetime.now().strftime('%Y%m%d')}-{randint(1000, 9999)}"


def create_order_from_cart(db: Session, user_id: int, payload: OrderCreate) -> Order:
    if payload.payment_method not in VALID_PAYMENT_METHODS:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid payment method")

    cart = cart_repository.get_or_create_cart(db, user_id)
    if not cart.items:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Cart cannot be empty")

    subtotal = 0.0
    order_items: list[OrderItem] = []
    for cart_item in cart.items:
        menu_item = menu_repository.get_item(db, cart_item.menu_item_id)
        if not menu_item or not menu_item.available:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Menu item {cart_item.menu_item_id} is not available")
        if menu_item.stock < cart_item.quantity:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=f"Not enough stock for {menu_item.name}")

        line_total = menu_item.price * cart_item.quantity
        subtotal += line_total
        order_items.append(
            OrderItem(
                menu_item_id=menu_item.id,
                item_name_snapshot=menu_item.name,
                item_price_snapshot=menu_item.price,
                quantity=cart_item.quantity,
                special_instructions=cart_item.special_instructions,
                line_total=line_total,
            )
        )
        menu_item.stock -= cart_item.quantity

    discount = round(subtotal * 0.10, 2) if payload.promo_code and payload.promo_code.upper() == "SAVE10" else 0.0
    total = round(subtotal - discount + DELIVERY_FEE, 2)
    payment_status = "Pending" if payload.payment_method == "CashOnDelivery" else "Paid"

    order = Order(
        user_id=user_id,
        order_number=_generate_order_number(),
        status="Pending",
        subtotal=round(subtotal, 2),
        discount=discount,
        delivery_fee=DELIVERY_FEE,
        total=total,
        delivery_address=payload.delivery_address,
        phone=payload.phone,
        payment_method=payload.payment_method,
        payment_status=payment_status,
    )

    created = order_repository.create_order(db, order, order_items)
    cart_repository.clear_cart(db, cart)
    return _order_response(created)


def list_orders(db: Session, user_id: int) -> list[Order]:
    return [_order_response(order) for order in order_repository.list_user_orders(db, user_id)]


def get_order(db: Session, user_id: int, order_id: int) -> Order:
    order = order_repository.get_user_order(db, user_id, order_id)
    if not order:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Order not found")
    return _order_response(order)


def cancel_order(db: Session, user_id: int, order_id: int) -> Order:
    order = get_order(db, user_id, order_id)
    if order.status not in CANCELABLE_STATUSES:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Order can no longer be cancelled")
    order.status = "Cancelled"
    return _order_response(order_repository.save_order(db, order))

File: src/services/orderService.ts
import { apiClient } from "./apiClient";
import type { CartItem, CartSummary } from "../types/cart";
import type { Order, OrderStatus, PaymentMethod } from "../types/order";
import type { User } from "../types/auth";
import { mapMenuItem, type BackendMenuItem } from "./menuService";

interface CheckoutPayload {
  user: User;
  items: CartItem[];
  summary: CartSummary;
  deliveryAddress: string;
  phone: string;
  paymentMethod: PaymentMethod;
}

interface BackendOrderItem {
  id: number;
  menu_item_id: number;
  item_name_snapshot: string;
  item_price_snapshot: number;
  quantity: number;
  special_instructions?: string | null;
  line_total: number;
}

interface BackendOrder {
  id: number;
  user_id: number;
  order_number: string;
  customer_name?: string | null;
  status: string;
  subtotal: number;
  discount: number;
  delivery_fee: number;
  total: number;
  delivery_address: string;
  phone: string;
  payment_method: string;
  created_at: string;
  items: BackendOrderItem[];
}

function toBackendPayment(method: PaymentMethod): string {
  return method === "Cash on Delivery" ? "CashOnDelivery" : method;
}

function fromBackendPayment(method: string): PaymentMethod {
  return method === "CashOnDelivery" ? "Cash on Delivery" : (method as PaymentMethod);
}

export function toBackendStatus(status: OrderStatus): string {
  return status === "Out for Delivery" ? "OutForDelivery" : status;
}

export function fromBackendStatus(status: string): OrderStatus {
  return status === "OutForDelivery" ? "Out for Delivery" : (status as OrderStatus);
}

function mapOrder(order: BackendOrder): Order {
  return {
    id: String(order.id),
    orderNumber: order.order_number,
    customerName: order.customer_name ?? "Customer",
    customerId: String(order.user_id),
    status: fromBackendStatus(order.status),
    deliveryAddress: order.delivery_address,
    phone: order.phone,
    paymentMethod: fromBackendPayment(order.payment_method),
    subtotal: order.subtotal,
    discount: order.discount,
    deliveryFee: order.delivery_fee,
    total: order.total,
    createdAt: order.created_at,
    estimatedMinutes: 45,
    items: order.items.map((item) => ({
      id: String(item.id),
      menuItem: mapMenuItem({
        id: item.menu_item_id,
        name: item.item_name_snapshot,
        category: "Pizza",
        description: item.special_instructions || "Ordered item",
        price: item.item_price_snapshot,
        image_url: null,
        stock: 0,
        available: true,
      } satisfies BackendMenuItem),
      quantity: item.quantity,
      price: item.item_price_snapshot,
      specialInstructions: item.special_instructions ?? undefined,
    })),
  };
}

export async function createOrder(payload: CheckoutPayload): Promise<Order> {
  const promoCode = localStorage.getItem("cos.promo") || undefined;
  const response = await apiClient.post<BackendOrder>("/api/orders", {
    delivery_address: payload.deliveryAddress,
    phone: payload.phone,
    payment_method: toBackendPayment(payload.paymentMethod),
    promo_code: promoCode,
  });
  localStorage.removeItem("cos.promo");
  return mapOrder(response);
}

export async function getCustomerOrders(): Promise<Order[]> {
  const response = await apiClient.get<BackendOrder[]>("/api/orders");
  return response.map(mapOrder);
}

export async function getOrder(id: string): Promise<Order | undefined> {
  return mapOrder(await apiClient.get<BackendOrder>(`/api/orders/${id}`));
}

export async function cancelOrder(id: string): Promise<void> {
  await apiClient.put<BackendOrder>(`/api/orders/${id}/cancel`);
}
